README.md
# Drafter PHP Wrapper
PHP wrapper for [Drafter](https://github.com/apiaryio/drafter) API Blueprint Parser harness.

[![Minimum PHP Version](https://img.shields.io/badge/php-%3E%3D%205.6-8892BF.svg)](https://php.net/)
[![Build Status](https://travis-ci.org/hendrikmaus/drafter-php.svg?branch=master)](https://travis-ci.org/hendrikmaus/drafter-php)
[![codecov.io](http://codecov.io/github/hendrikmaus/drafter-php/coverage.svg?branch=master)](http://codecov.io/github/hendrikmaus/drafter-php?branch=master)
[![Dependency Status](https://www.versioneye.com/user/projects/55e7ed98211c6b00190007d2/badge.svg?style=flat)](https://www.versioneye.com/user/projects/55e7ed98211c6b00190007d2) 
[![Code Climate](https://img.shields.io/codeclimate/github/kabisaict/flow.svg)](https://codeclimate.com/github/hendrikmaus/drafter-php)

## What is Drafter-php?
Drafter-php allows you to use use the [drafter](https://github.com/apiaryio/drafter) API Blueprint Parser harness
with your PHP application.

In a nutshell: you can convert [API Blueprint](http://apiblueprint.org/) files to parse result.

[API Blueprint](http://apiblueprint.org/) is a webservice documentation language built on top of 
[Markdown](https://en.wikipedia.org/wiki/Markdown).

## Requirements
* PHP 5.6 or greater
* Drafter [command line tool](https://github.com/apiaryio/drafter#drafter-command-line-tool)

## What Is What
**Drafter** is a C++ tool to parse API Blueprint.  
**Drafter-php** is a PHP wrapper around the **Drafter command line tool**.

## Installation
The recommended way to install Drafter-php is by using [composer](https://getcomposer.org):

```bash
$ composer require hmaus/drafter-php
```

This will install the PHP package with your application.  
Please keep in mind that **Drafter is not included**.

### Install Drafter Command Line Tool using Composer
Head over to [hmaus/drafter-installer](https://github.com/hendrikmaus/drafter-installer).

## Usage of Drafter-php
1. Get an instance of the `\DrafterPhp\DrafterInterface` implementation, `\DrafterPhp\Drafter`
1.1 You will need to pass the path to your drafter binary to the constructor
1.2 It is recommended to solve this using a dependency injection container
2. Set the input file and options on your `\DrafterPhp\Drafter` instance
2.1 Drafter-php currently does not support passing blueprint code directly to Drafter;
    it has to be stored in a file at this time
3. Run your command
    
### Input / Output Examples

> Note: drafter-php does not assert the structure of the output.
> If you see differences in the examples to your actual output, please refer to the official drafter docs.

Given this api blueprint source:

```apib
# GET /message
+ Response 200 (text/plain)

        Hello World!
```

The result will look similar (json refract):

```json
{
  "element": "parseResult",
  "content": [
    {
      "element": "category",
      "meta": {
        "classes": [
          "api"
        ],
        "title": ""
      },
      "content": [
        {
          "element": "category",
          "meta": {
            "classes": [
              "resourceGroup"
            ],
            "title": ""
          },
          "content": [
            {
              "element": "resource",
              "meta": {
                "title": ""
              },
              "attributes": {
                "href": "/message"
              },
              "content": [
                {
                  "element": "transition",
                  "meta": {
                    "title": ""
                  },
                  "content": [
                    {
                      "element": "httpTransaction",
                      "content": [
                        {
                          "element": "httpRequest",
                          "attributes": {
                            "method": "GET"
                          },
                          "content": []
                        },
                        {
                          "element": "httpResponse",
                          "attributes": {
                            "statusCode": "200",
                            "headers": {
                              "element": "httpHeaders",
                              "content": [
                                {
                                  "element": "member",
                                  "content": {
                                    "key": {
                                      "element": "string",
                                      "content": "Content-Type"
                                    },
                                    "value": {
                                      "element": "string",
                                      "content": "text/plain"
                                    }
                                  }
                                }
                              ]
                            }
                          },
                          "content": [
                            {
                              "element": "asset",
                              "meta": {
                                "classes": [
                                  "messageBody"
                                ]
                              },
                              "attributes": {
                                "contentType": "text/plain"
                              },
                              "content": "Hello World!\n"
                            }
                          ]
                        }
                      ]
                    }
                  ]
                }
              ]
            }
          ]
        }
      ]
    }
  ]
}
```

### Code Examples

> Found something wrong? Feel free to [contribute](https://github.com/hendrikmaus/drafter-php/blob/master/CONTRIBUTING.md)

#### Make sure it works
To make sure it works, we'll ask Drafter for the current version.

```php
$version = $drafter
    ->version()
    ->run();

// Reset options on the command
$drafter->resetOptions();
```
`$version` should now contain a string like `v1.0.0`.
 If something is wrong, an exception will have been thrown most likely.
 
 > Keep in mind that Drafter-php is designed to keep its state,
 > run `\DrafterPhp\DrafterInterface::resetOptions` to get rid of the version option you just set
 > for the next call on the instance.

#### Parse your-service.apib into your-service.refract.json
Make sure your input path is correct and readable, and your output path is writable.

```php
$drafter
    ->input('your-service.apib')
    ->format('json')
    ->type('refract')
    ->output('your-service.refract.json')
    ->run();
```

#### Parse your-service.apib into your-service.ast.json
Make sure your input path is correct and readable, and your output path is writable.

```php
$drafter
    ->input('your-service.apib')
    ->format('json')
    ->output('your-service.ast.json')
    ->run();
```

#### Parse your-service.apib into a PHP data structure

```php
$refract = $drafter
    ->input('your-service.apib')
    ->format('json')
    ->run();
    
$phpObj = json_decode($refract);
$phpArr = json_decode($refract, true);
```

#### Parse your-service.apib into YAML format

```php
$drafter
    ->input('your-service.apib')
    ->format('yaml') // optional as yaml is the default
    ->output('your-service.ast.yml')
    ->run();
```

#### Get Process before it is run

```php
$process = $drafter
    ->input('your-service.apib')
    ->format('json')
    ->output('your-service.refract.json')
    ->build();

// do stuff with the process

$drafter
    ->run($process);
```

## Feature Roadmap
Do not hesitate to [contribute](https://github.com/hendrikmaus/drafter-php/blob/master/CONTRIBUTING.md).

* [ ] support passing raw api blueprint code into `\DrafterPhp\DrafterInterface::input`, rather than always a file path

## License
Drafter-php is licensed under the MIT License - see the 
[LICENSE](https://github.com/hendrikmaus/drafter-php/blob/master/LICENSE) file for details


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

The change log itself is written the way that [keepachangelog.com](http://keepachangelog.com/) describes.

## [Unreleased]

## [6.0.0] - 2019-11-24
## Changed
- Update drafter dependency to v4.0.2 [BC]
- Removed `type` option

## [5.0.1] - 2018-07-20
## Fixed
- Allow spaces in filenames

## [5.0.0] - 2018-03-05
## Changed
- Allow v4 of Symfony component
- Update Robo task runner

## [4.0.1] - 2016-09-20
## Changed
- Removed the cumbersome way to install drafter and added a reference to hmaus/drafter-installer
- Renamed phpunit.xml to phpunit.xml.dist and ignored phpunit.xml so you can customize your settings
- Slimmed down phpunit output

## [4.0.0] - 2016-08-05
## Changed
- Update drafter dependency to v3.0.0 [BC]

## [3.0.0] - 2016-03-31
## Changed
- Updated drafter dependency to v2.2.0 in composer.json and readme [BC]
- Implemented new behavior of the sourcemap argument [BC]
- Begin namespace with vendor name `Hmaus` [BC]

## [2.0.0] - 2015-11-07
## Changed
- Increased min required PHP version to >=5.6 [BC]
- Bumped PHPUnit to v5.x
- Composer install example always uses the latest master of drafter
- Allow v3 of Symfony component

## Removed
- Tests do not care about the exact drafter output
- Exact match fixtures used by the removed test assertions

## [1.0.0] - 2015-10-04
## Changed
- Install example to drafter v1.0.0

## Added
- Support for drafters new `type` option which defaults to `refract`

### Removed
- `symfony/console`
- `symfony/var-dumper` (--dev)

## [0.1.0] - 2015-09-03
### Added
- Initial version


tests/DrafterTest.php
<?php
/**
 * Test drafter binary wrapper.
 *
 * todo, test:
 *  - output cannot be written to disk
 *  - validate options is set, validation fails
 *  - use line num behavior
 *
 * @author    hmaus
 * @since     2015-08-28
 * @copyright 2015 (c) Hendrik Maus
 * @package   DrafterPhp
 */

namespace Hmaus\DrafterPhp\Tests;

use Hmaus\DrafterPhp\Drafter;
use Hmaus\DrafterPhp\DrafterInterface;
use Symfony\Component\Process\Process;

class DrafterTest extends \PHPUnit_Framework_TestCase
{
    /**
     * Path to the drafter bin for the tests.
     * @var string
     */
    private $binPath;

    /**
     * Base path for apib fixtures for the tests.
     * @var string
     */
    private $fixturePath;

    /**
     * Drafter php instance.
     * @var DrafterInterface
     */
    private $drafter;

    /**
     * Init test case.
     */
    protected function setUp()
    {
        $this->binPath     = __DIR__ . '/../vendor/bin/drafter';
        $this->fixturePath = __DIR__ . '/fixtures/';
        $this->drafter     = new Drafter($this->binPath);
    }

    /**
     * Create a drafter instance; check if it worked.
     */
    public function testCreateWrapperObject()
    {
        $this->assertInstanceOf(Drafter::class, $this->drafter);
    }

    /**
     * Test configuring the path to drafter binary.
     */
    public function testSettingAndGettingTheBinaryPath()
    {
        $this->assertEquals($this->binPath, $this->drafter->getBinary());
        $this->drafter->setBinary('new.path');
        $this->assertEquals('new.path', $this->drafter->getBinary());
    }

    /**
     * Get drafter binary version.
     */
    public function testGetVersion()
    {
        $version = $this->drafter->version()->run();

        // the drafter binary will add a line break at the end of the version string
        $version = trim($version);

        $this->assertNotEmpty($version);
    }

    /**
     * Parse a simple example to json ast.
     */
    public function testParseSimplestExampleToJson()
    {
        $json = $this
            ->drafter
            ->input($this->fixturePath . 'simplest-example.apib')
            ->format('json')
            ->run();

        $this->assertNotNull($json);
        $this->assertJson($json);

        return $this->drafter;
    }

    /**
     * The drafter instance will hold its state; we can invoke `run` again for the same result.
     *
     * @param Drafter $drafter
     * @depends testParseSimplestExampleToJson
     * @throws \Exception
     */
    public function testRunningDrafterAgainWithoutReset(Drafter $drafter)
    {
        $this->assertJson($drafter->run());
    }

    /**
     * Build the process instance using drafter and pass it to the run method.
     */
    public function testBuildProcessAndPassItToRunMethod()
    {
        $process = $this
            ->drafter
            ->input($this->fixturePath . 'simplest-example.apib')
            ->format('json')
            ->build();

        $this->assertInstanceOf(Process::class, $process);

        $json = $this
            ->drafter
            ->run($process);

        $this->assertJson($json);
    }

    /**
     * Build the process instance using drafter but with space in file name (input and output)
     * @expectedException \Exception
     * @expectedExceptionMessageRegExp /unable to open file/
     */
    public function testBuildWithSpace()
    {
        $process = $this
            ->drafter
            ->input($this->fixturePath . 'simplest example.apib')
            ->output($this->fixturePath . 'simplest example.json')
            ->format('json')
            ->build();

        $this->assertInstanceOf(Process::class, $process);

        $this
            ->drafter
            ->run($process);
    }

    /**
     * Test setting all supported options with correct syntax.
     */
    public function testSetAllAvailableOptions()
    {
        $this
            ->drafter
            ->input('some-input')
            ->output('some-output')
            ->version()
            ->validate()
            ->format('json')
            ->sourcemap()
            ->useLineNum();

        $expected = [
            '--output'       => 'some-output',
            '--version'      => '',
            '--validate'     => '',
            '--format'       => 'json',
            '--sourcemap'    => '',
            '--use-line-num' => ''
        ];

        $options = $this->drafter->getOptions();

        $this->assertEquals($expected, $options);
    }

    /**
     * Test resetting the input argument.
     */
    public function testResetInputArgument()
    {
        $this
            ->drafter
            ->input('some-input');

        $this->assertEquals('some-input', $this->drafter->getInput());

        $this
            ->drafter
            ->resetInput();

        $this->assertNull($this->drafter->getInput());
    }

    /**
     * Test resetting all options on the instance.
     */
    public function testResetOptions()
    {
        $this
            ->drafter
            ->version();

        $this->assertCount(1, $this->drafter->getOptions());

        $this
            ->drafter
            ->resetOptions();

        $this->assertCount(0, $this->drafter->getOptions());
    }

    /**
     * Catch an exception if the binary cannot be found.
     *
     * @expectedException \Exception
     * @expectedExceptionMessageRegExp /INVALID/
     */
    public function testInvalidBinaryWillThrowException()
    {
        $this
            ->drafter
            ->setBinary('INVALID')
            ->version()
            ->run();
    }

    /**
     * Catch an exception if the input file cannot be opened.
     *
     * @expectedException \Exception
     * @expectedExceptionMessage fatal: unable to open file 'INVALID'
     */
    public function testInvalidInputArgumentWillThrowException()
    {
        $this
            ->drafter
            ->input('INVALID')
            ->run();
    }

    /**
     * Catch an exception if an invalid format is passed.
     *
     * @expectedException \Exception
     * @expectedExceptionMessage option value is invalid: --format=INVALID
     */
    public function testInvalidFormatWillThrowException()
    {
        $this
            ->drafter
            ->input($this->fixturePath . 'simplest-example.apib')
            ->format('INVALID')
            ->run();
    }

    /**
     * Catch an exception if no option or argument is set.
     *
     * @expectedException \Exception
     * @expectedExceptionMessage Input argument missing
     */
    public function testExceptionWhenInputArgumentIsMissing()
    {
        $this->drafter->run();
    }
}


src/DrafterInterface.php
<?php
/**
 * @author    hmaus
 * @since     2015-08-28
 * @copyright 2015 (c) Hendrik Maus
 * @package   DrafterPhp
 */

namespace Hmaus\DrafterPhp;

use Symfony\Component\Process\Process;

interface DrafterInterface
{
    /**
     * Set path to process by Drafter.
     *
     * @param string $path input apib file path, e.g. /your-service/docs/your-service.apib
     * @return $this
     */
    public function input($path);

    /**
     * Set output path.
     *
     * Drafter will write the abstract syntax tree to this location.
     * Make sure it is writable beforehand.
     *
     * Help:
     *   -o, --output save output AST into file (string [=])
     *
     * @param string $path output path
     * @return $this
     */
    public function output($path);

    /**
     * Set format option.
     *
     * Available formats:
     *   * yaml (default)
     *   * json
     *
     * Help:
     *   -f, --format output AST format (string [=yaml])
     *
     * @param string $format output format to set, e.g. 'json'
     * @return $this
     */
    public function format($format);

    /**
     * Set sourcemap argument.
     *
     * Drafter will export sourcemap in the Parse Result
     *
     * Help:
     *   -s, --sourcemap export sourcemap in the Parse Result
     *
     * @return $this
     */
    public function sourcemap();

    /**
     * Get drafter version.
     *
     * Note: Drafter will add a line break to the end of the string,
     * the return value will actually be "v0.1.9\n"
     *
     * Help:
     *   -v, --version print Drafter version
     *
     * @return $this
     */
    public function version();

    /**
     * Set validate option.
     *
     * Help:
     *   -l, --validate validate input only, do not print AST
     *
     * @return $this
     */
    public function validate();

    /**
     * Set use line number option.
     *
     * Help:
     *   -u, --use-line-num use line and row number instead of character index when printing annotation
     *
     * @return $this
     */
    public function useLineNum();

    /**
     * Build the process and return it without running it.
     *
     * The method will be called by `\DrafterPhp\DrafterInterface::run`
     * You can use it, to inspect the process before it is run.
     *
     * To run the process, pass it to `\DrafterPhp\DrafterInterface::run`,
     * the method will do exception handling for you and will return the result.
     *
     * @return Process
     */
    public function build();

    /**
     * Run Drafter command.
     *
     * todo the return value of this method behaves differently depending on the options passed to drafter
     *  i.e. if we do not set the --output option, the method will return the result as a string;
     *  if we provide it, the method will return null
     *  this needs to be taken into account for tests and documentation.
     *
     * @return string
     * @param Process $process (optional) the process obj if you fetched it from `\DrafterPhp\DrafterInterface::build`
     *                         Otherwise, the method will build the process on its own using the aforementioned method.
     * @throws \Exception
     */
    public function run(Process $process = null);

    /**
     * Get the raw associative options array.
     *
     * E.g.:
     * ```php
     * ['--output' => 'path', '--format' => 'json']
     * ```
     *
     * @return string[]
     */
    public function getOptions();

    /**
     * Reset the argument list to not keep any state.
     *
     * @return $this;
     */
    public function resetOptions();

    /**
     * Get the binary path that was set.
     *
     * @return string
     */
    public function getBinary();

    /**
     * Set the path to the Drafter binary.
     * This method will neither check if the binary exists, nor will it do any other validation.
     *
     * @param string $binary
     * @return $this
     */
    public function setBinary($binary);

    /**
     * Get input file path.
     *
     * @return string[]
     */
    public function getInput();

    /**
     * Reset input file path.
     *
     * @return $this
     */
    public function resetInput();
}


src/Drafter.php
<?php
/**
 * Access to the drafter binary.
 *
 * This component will provide a PHP interface to the drafter binary.
 *
 * ```php
 * $ast = $drafter
 *  ->input('blueprint.apib')
 *  ->format('json')
 *  ->output('ast.json')
 *  ->run();
 * ```
 *
 * Once drafter is configured, the object instance will not reset its
 * arguments and options.
 * So running `$drafter->run()` again immediately, will return the very same result.
 *
 * This behaviour allows you to re-use a drafter instance without minimal config change per call.
 *
 * More about the internals of this wrapper:
 * Output of `drafter --help`:
 *
 * ```
 * usage: drafter [options] ... <input file>
 *
 * API Blueprint Parser
 * If called without <input file>, 'drafter' will listen on stdin.
 *
 * options:
 * -o, --output          save output Parse Result into file (string [=])
 * -f, --format          output format of the Parse Result (yaml|json) (string [=yaml])
 * -t, --type            type of the Parse Result (refract|ast) (string [=refract])
 * -s, --sourcemap       export sourcemap in the Parse Result
 * -h, --help            display this help message
 * -v, --version         print Drafter version
 * -l, --validate        validate input only, do not output Parse Result
 * -u, --use-line-num    use line and row number instead of character index when printing annotation
 * ```
 *
 * @author    hmaus
 * @since     2015-08-28
 * @copyright 2015 (c) Hendrik Maus
 * @package   DrafterPhp
 */

namespace Hmaus\DrafterPhp;

use Symfony\Component\Process\Process;

class Drafter implements DrafterInterface
{
    /**
     * @var string
     */
    private $binary;

    /**
     * Collect options, e.g. output, format etc.
     *
     * Format:
     *   '--option-name' => 'option value'
     *   '--output'      => 'path/to/output/file'
     *
     * @var string[]
     */
    private $options = [];

    /**
     * Input file path to pass into drafter as argument.
     *
     * @var string
     */
    private $input;

    /**
     * Create a drafter wrapper by passing the path to your drafter bin.
     *
     * Hint:
     *  You can usually reflect this with your dependency injection container.
     *
     * @param string $binPath Path to the drafter binary
     */
    public function __construct($binPath)
    {
        $this->binary = $binPath;
    }

    public function input($path)
    {
        $this->input = $path;

        return $this;
    }

    public function output($path)
    {
        $this->options['--output'] = $path;

        return $this;
    }

    public function format($format)
    {
        $this->options['--format'] = $format;

        return $this;
    }

    public function sourcemap()
    {
        $this->options['--sourcemap'] = '';

        return $this;
    }

    public function version()
    {
        $this->options['--version'] = '';

        return $this;
    }

    public function validate()
    {
        $this->options['--validate'] = '';

        return $this;
    }

    public function useLineNum()
    {
        $this->options['--use-line-num'] = '';

        return $this;
    }

    public function build()
    {
        $this->validateOptionsAndArguments();

        $process = new Process(
            $this->getProcessCommand()
        );

        return $process;
    }

    public function run(Process $process = null)
    {
        if (null === $process) {
            $process = $this->build();
        }

        $process->run();

        if ($process->getExitCode() !== 0) {
            throw new \Exception($process->getErrorOutput());
        }

        return $process->getOutput();
    }

    public function getOptions()
    {
        return $this->options;
    }

    public function resetOptions()
    {
        $this->options = [];

        return $this;
    }

    public function getBinary()
    {
        return $this->binary;
    }

    public function setBinary($binary)
    {
        $this->binary = $binary;

        return $this;
    }

    public function getInput()
    {
        return $this->input;
    }

    public function resetInput()
    {
        $this->input = null;

        return $this;
    }

    /**
     * Check the configuration of the object prior to executing the process.
     *
     * As we do not support listening to stdin, drafter must not be called
     * without an input argument.
     *
     * If the version is requested, we do not require
     * the input argument as it will be ignored by drafter anyway.
     *
     * @throws \Exception
     */
    private function validateOptionsAndArguments()
    {
        if ($this->isVersionOptionSet()) {
            return;
        }

        if (!$this->input) {
            throw new \Exception("Input argument missing");
        }
    }

    /**
     * Helper to find out if the --version option was set.
     *
     * @return bool
     */
    private function isVersionOptionSet()
    {
        return array_key_exists('--version', $this->options);
    }

    /**
     * Get command to pass it into the process.
     *
     * The method will append the input file path argument to the end, like
     * described in the `drafter --help` output.
     *
     * @return string
     */
    private function getProcessCommand()
    {
        $options   = $this->transformOptions();
        $options[] = escapeshellarg($this->input);

        $command = escapeshellarg($this->binary) . ' ' . implode(' ', $options);

        return $command;
    }

    /**
     * Return options prepared to be passed into the Process.
     *
     * E.g.: ["--output" => "path/to/file"] -> ["--output=path/to/file"]
     *
     * The original format is an associative array, where the key is the
     * option name and the value is the respective value.
     * The process will want those as single strings to escape them
     * for the command line. Hence, we have to turn ["--output" => "path/to/file"]
     * into ["--output=path/to/file"].
     *
     * @return \string[]
     */
    private function transformOptions()
    {
        $processOptions = [];

        foreach ($this->options as $key => $value) {
            $option = $key;

            if ($value) {
                $option .= '=' . escapeshellarg($value);
            }

            $processOptions[] = $option;
        }

        return $processOptions;
    }
}
